tests/projection_test.py

```diff
@@ -222,6 +234,21 @@ def test_projection_hyperplane(self):
     p = projection.projection_hyperplane(x, (a, b))
     self.assertAllClose(jnp.dot(a, p), b)
 
+    tree_x = (1.0, {"k1": 2.0, "k2": (2.0, 1.0)}, 2.0)
+    tree_a = (1.0, {"k1": 1.0, "k2": (1.0, 0.0)}, 1.0)
+    p = projection.projection_hyperplane(tree_x, (tree_a, b))
+    expected_p = (-0.5, {"k1": 0.5, "k2": (0.5, 1.0)}, 0.5)
+    self.assertAllClose(expected_p, p)
+
+  def test_projection_halfspace(self):
+    # Testing that for b large enough, halfplane projection of x is x.
+    rng = onp.random.RandomState(0)
+    tree_x = (rng.randn(1), {"k1": rng.randn(1), "k2": (rng.randn(1), rng.randn(1))}, rng.randn(1))
+    tree_a = (rng.randn(1), {"k1": rng.randn(1), "k2": (rng.randn(1), rng.randn(1))}, rng.randn(1))
+    b = 10000
+    p = projection.projection_halfspace(tree_x, (tree_a, b))
+    self.assertAllClose(tree_x, p)
```

Comment:
Thanks for creating this test! I hadn't realized that we were missing it. 

Could you also check the case when `b` is small? A simple assertion is to just check that the solution is feasible. A more advanced assertion would be to check that the optimality conditions are satisfied.